pennylane/devices/null_qubit.py

```diff
@@ -0,0 +1,284 @@
+        self._operation_calls["PauliY"] += 1
+
+    def _apply_z(self, state, axes, **kwargs):
+        self._operation_calls["PauliZ"] += 1
+
+    def _apply_hadamard(self, state, axes, **kwargs):
+        self._operation_calls["Hadamard"] += 1
+
+    def _apply_s(self, state, axes, inverse=False):
+        self._operation_calls["S"] += 1
+
+    def _apply_t(self, state, axes, inverse=False):
+        self._operation_calls["T"] += 1
+
+    def _apply_sx(self, state, axes, inverse=False):
+        self._operation_calls["SX"] += 1
+
+    def _apply_cnot(self, state, axes, **kwargs):
+        self._operation_calls["CNOT"] += 1
+
+    def _apply_swap(self, state, axes, **kwargs):
+        self._operation_calls["SWAP"] += 1
+
+    def _apply_cz(self, state, axes, **kwargs):
+        self._operation_calls["CZ"] += 1
+
+    def _apply_toffoli(self, state, axes, **kwargs):
+        self._operation_calls["Toffoli"] += 1
+
+    def _apply_phase(self, state, axes, parameters, inverse=False):
+        pass
+
+    def expval(self, observable, shot_range=None, bin_size=None):
+        pass
+
+    def var(self, observable, shot_range=None, bin_size=None):
+        pass
+
+    @classmethod
+    def capabilities(cls):
+        capabilities = super().capabilities().copy()
+        capabilities.update(
+            model="qubit",
+            supports_inverse_operations=True,
+            supports_analytic_computation=True,
+            supports_broadcasting=True,
+            returns_state=True,
+            passthru_devices={
+                "tf": "null.qubit.tf",
+                "torch": "null.qubit.torch",
+                "autograd": "null.qubit.autograd",
+                "jax": "null.qubit.jax",
+            },
+        )
+        return capabilities
+
+    @staticmethod
+    def _create_basis_state(index):
+        return None
+
+    @property
+    def state(self):
+        return None
+
+    def density_matrix(self, wires):
+        return None
+
+    def _apply_state_vector(self, state, device_wires):
+        pass
+
+    def _apply_basis_state(self, state, wires):
+        pass
+
+    def _apply_unitary(self, state, mat, wires):
+        pass
+
+    def _apply_unitary_einsum(self, state, mat, wires):
+        pass
+
+    def _apply_diagonal_unitary(self, state, phases, wires):
+        pass
+
+    def reset(self):
+        pass
+
+    def analytic_probability(self, wires=None):
+        pass
+
+    def generate_samples(self):
+        """Returns the computational basis samples generated for all wires.
+        In the _qubit_device.py, the function calls for analytic_probability for its operations."""
+        self.analytic_probability()
+
+    def sample(self, observable, shot_range=None, bin_size=None, counts=False):
+        pass
+
+    def operation_calls(self):
+        """Statistics of operation calls"""
+        return self._operation_calls
+
+    def execute(self, circuit, **kwargs):
+        self.apply(circuit.operations)
```

Comment:
Hi @antalszava, could you please elaborate more on that?

Reply:
Sure :) In [`QubitDevice.execute`](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/_qubit_device.py#L315), in addition to the `operations`, we also apply the diagonalizing gates of the tape:
```py
        self.apply(circuit.operations, rotations=circuit.diagonalizing_gates, **kwargs)
```
In addition, we have logic that computes the statistics for the specified observables,
```py
        # compute the required statistics
        if not self.analytic and self._shot_vector is not None:
            results = self._collect_shotvector_results(circuit, counts_exist)
        else:
            results = self.statistics(circuit.observables, circuit=circuit)
```
I was just curious if these should be considered in some way here.

Reply:
Hi @antalszava, thank you for your suggestion.
I expanded the `Nullqubit` apply method definitions to accept rotations and this integrated nicely into the suite of tests.
Regarding the logic for the computation of statistics, I have no plan of expanding the execute method to include that, mainly because we don't want to compute statistics.

Reply:
Makes sense, thanks :)